src/io/json/write/serialize.rs

```diff
@@ -299,6 +298,17 @@ fn set_column_for_json_rows(
                 .for_each(|(row, obj)| {
```

Comment:
this `for_each` will hit the iterator, causing a complete serialization even if the item is `null` from the struct's point of view.

what about something like `if let Some(validity) = array.validity()`, we `zip` the validity in this iteration directly (and if not, we use the code we already have here). I.e. essentially try to "blend" your changes to the existing code so that we avoid running the loop twice.